File: kotlin/java-basic-example-vs-kotlin/src/main/java/org/example/ex03/Exam03.java
package org.example.ex03;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;

import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

/**
 * @author : Rene
 * @since : 2023/10/22
 */

@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
class User {
	private String name;
	private int age;
}

public class Exam03 {

	public Exam03() {
		var userList = new ArrayList<User>();
		userList.add(new User("1", 10));
		userList.add(new User("2", 10));
		userList.add(new User("3", 10));

		var list = Arrays.asList(
			new User("4" , 20),
			new User("5" , 20),
			new User("6" , 20)
		);

		userList.forEach(System.out::println);

		for (int i =0; i< userList.size(); i++){
			var dto = userList.get(i);
			System.out.println("i = " + i);
		}


		// 자바에서도 immutable을 제공 -> but 가변 메서드가 사용되기 때문에 -> 에러 발생
		var immutable = Collections.unmodifiableCollection(userList);
		immutable.add(new User("5", 60));

	}

}

File: kotlin/kotlin-basic-example/src/main/kotlin/ex03/Exam03.kt
package ex03

import java.util.Collections

/**
 * @author        : Rene
 * @since         : 2023/10/22
 */
fun main() {
    // 코틀린에서는 불변과 가변을 중요하게 나눔
    // 변수에서처럼 컬렉션에서도 ->

    val userList = mutableListOf<User>()
    userList.add(User("1", 10))
    userList.add(User("2", 10))
    userList.add(User("3", 10))

    // 불변 -> 반드시 초기화때 넣어주어야 함 -> add라는 메서드가 지원을 안함
    // 자바에서는 불변이어도 add가 보이지만
    val list = listOf<User>(
        User("4", 40)
    )

    for(element in userList){
        println(element)
    }

    userList.forEach{it->println(it)}
    userList.forEach{ println(it)}

    // index를 보고 싶은 경우에
    userList.forEachIndexed{index, user -> println("$index, $user") }
    userList.forEachIndexed(fun (index, user){
        println("$index, $user")
    })


    // for문에다가 넣는 방식
    for ((index, element) in userList.withIndex()){
        println("$index, $element")
    }
}

class User(
    var name: String,
    var age: Int
)